posts/frangeris/03.secure-values-by-private-keys-in-aws.md

```diff
@@ -0,0 +1,235 @@
+
+> This implementation focuses on using specifics services and functionalities provided by AWS services. Implementations will vary depending on the provider (Azure, GCP).
+
+### Let's start by creating a basic [AWS Lambda](https://aws.amazon.com/lambda/)
+
+This function will try to access our "secure" and use it. By using `npx` command we get access to serverless executable, select the basic to keep it simple, there are options for different types of uses cases, even different languages options in case `Node.js` is not your language:
+
+![Run npx serverless command to start](./assets/03.image.1.png)
+
+Some operations used on the lambda function will require some permissions later, due to this, it's recommended to include and enable those permissions before the IAM role of the lambda gets created, here is what permissions we need the lambda to have:
+
+```yaml
+provider:
+  # ...
+  iam:
+    role:
+      statements:
+        - Effect: Allow
+          Action:
+            - "s3:*"
+          Resource: "*"
+        - Effect: Allow
+          Action:
+            - "cloudwatch:*"
+            - "logs:*"
+          Resource: "*"
+        - Effect: Allow
+          Action:
+            - "ssm:GetParameter"
+            - "ssm:GetParametersByPath"
+          Resource: "*"
+        - Effect: Allow
+          Action:
+            - "secretsmanager:GetSecretValue"
+          Resource: "*"
+        - Effect: Allow
+          Action:
+            - "kms:Decrypt"
+          Resource: "*"
+```
+
+Then, with your [aws account](https://docs.aws.amazon.com/cli/latest/userguide/cli-configure-files.html) already set up, let's deploy it.
+
+```bash
+npx sls deploy
+```
+
+🎉 After a correct deployment, the code should be created successfully and ready to use, now we need to configure the most important part, security.
+
+## Security
+
+Permission will be managed via [IAM policies](https://docs.aws.amazon.com/IAM/latest/UserGuide/access_policies.html) with the objective of _only allows the lambda to get and decrypt the secured value_, the approach used will be based on: **prevent IAM entities from accessing the KMS key and allow the root user account to manage it** (also preventing the root user account from losing access to the KMS key).
+
+This "protection" layer of our sensitive value is based on the composition of two AWS services, [Key Management Service](https://aws.amazon.com/kms/) is the one responsible for creating the private key which will be used to encrypt our value, then [Secret Manager](https://aws.amazon.com/es/secrets-manager/) or [System Manager: Parameters Store](https://aws.amazon.com/systems-manager/), both services allows us to accomplish the same functionality but differ on some caveats, _the chosen one_ will be used for saving the secured encrypted value, here is a quick graphic representation of the implementation:
```

Comment:
```suggestion
This "protection" layer of our sensitive value is based on the composition of two AWS services: [Key Management Service](https://aws.amazon.com/kms/) is the one responsible for creating the private key which will be used to encrypt our value, then [Secret Manager](https://aws.amazon.com/es/secrets-manager/) or [System Manager: Parameters Store](https://aws.amazon.com/systems-manager/) allow us to accomplish the same functionality but they differ on some caveats, since _the chosen one_ will be used for saving the secured encrypted value. Here is a quick graphic representation of the implementation:
```